runtime/include/atomics/standard/chpl-atomics.h

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright 2004-2016 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef _chpl_atomics_h_
+#define _chpl_atomics_h_
+
+#include "chpltypes.h"
+
+#if defined(__cplusplus)
+  #if __cplusplus >= 201103L
+    #include <atomic>
+    #define Atomic(T) std::atomic<T>
+    using std::atomic_int_least8_t;
+    using std::atomic_int_least16_t;
+    using std::atomic_int_least32_t;
+    using std::atomic_int_least64_t;
+    using std::atomic_uint_least8_t;
+    using std::atomic_uint_least16_t;
+    using std::atomic_uint_least32_t;
+    using std::atomic_uint_least64_t;
+    using std::atomic_uintptr_t;
+    using std::atomic_bool;
+    using std::memory_order_relaxed;
+    using std::memory_order_consume;
+    using std::memory_order_acquire;
+    using std::memory_order_release;
+    using std::memory_order_acq_rel;
+    using std::memory_order_seq_cst;
+    using std::memory_order;
+  #else
+    #error "Standard atomics need at least C++11.  Use intrinsics or locks."
+  #endif
+#elif __STDC_VERSION__ >= 201112L && !defined(__STDC_NO_ATOMICS__)
```

Comment:
There is no equivalent macro for C++ because atomics are required for C++, and optional for C.

If the compiler lies about what it supports, there is nothing we can do about it.  The combination of macros is the best we can do.  For compilers that claim to support atomics but do not, the error message will just be less helpful because it will be about a missing stdatomic.h.

However, if a compiler claims not to support atomics, we sure don't want to ignore that, so we do need to check the feature-test macros.